File: L04_CleanCode/DiscountManager.cs
using System;

public class DiscountManager
{
    public static decimal ApplyDiscount(decimal price, AccountStatus accountStatus, int timeOfHavingAccountInYears)
    {
        decimal priceAfterDiscount = 0;
        switch (accountStatus)
        {
            case AccountStatus.NotRegistered:
                priceAfterDiscount = price;
                break;
            case AccountStatus.SimpleCustomer:
                priceAfterDiscount = ApplyDiscountForAccountStatus(price, DiscountForSimpleCustomers);
                break;
            case AccountStatus.ValuableCustomer:
                priceAfterDiscount = ApplyDiscountForAccountStatus(price, DiscountForValuableCustomers);
                break;
            case AccountStatus.MostValuableCustomer:
                priceAfterDiscount = ApplyDiscountForAccountStatus(price, DiscountForMostValuableCustomers);
                break;
            default:
                throw new NotImplementedException();
        }
        priceAfterDiscount = ApplyDiscountForTimeOfHavingAccount(priceAfterDiscount, timeOfHavingAccountInYears);
        return priceAfterDiscount;
    }

    public enum AccountStatus
    {
        NotRegistered,
        SimpleCustomer,
        ValuableCustomer,
        MostValuableCustomer
    }

    public const decimal DiscountForSimpleCustomers = 0.1m;
    public const decimal DiscountForValuableCustomers = 0.3m;
    public const decimal DiscountForMostValuableCustomers = 0.5m;

    public const int MaximumDiscountForLoyality = 5;

    public static decimal ApplyDiscountForAccountStatus(decimal price, decimal discountRatio)
    {
        return price - (discountRatio * price);
    }
    
    public static decimal ApplyDiscountForTimeOfHavingAccount(decimal price, int timeOfHavingAccountInYears)
    {
        decimal discountForLoyaltyInPercentage = (timeOfHavingAccountInYears > MaximumDiscountForLoyality) ? (decimal)MaximumDiscountForLoyality/100 : (decimal)timeOfHavingAccountInYears/100;
        return price - (discountForLoyaltyInPercentage * price);
    }
}

File: L04_CleanCode/TypeScript/DiscountManager.ts
enum AccountStatus
{
    NotRegistered,
    SimpleCustomer,
    ValuableCustomer,
    MostValuableCustomer
}

class DiscountManager
{
    public static readonly DiscountForSimpleCustomers: number = 0.1;
    public static readonly DiscountForValuableCustomers: number = 0.3;
    public static readonly DiscountForMostValuableCustomers: number = 0.5;

    public static readonly MaximumDiscountForLoyality: number = 5;

    public static ApplyDiscountForAccountStatus(price: number, discountRatio: number): number
    {
        return price - (discountRatio * price);
    }

    public static ApplyDiscountForTimeOfHavingAccount(price: number,timeOfHavingAccountInYears: number): number
    {
        let discountForLoyaltyInPercentage: number = (timeOfHavingAccountInYears > DiscountManager.MaximumDiscountForLoyality) ? DiscountManager.MaximumDiscountForLoyality/100 : timeOfHavingAccountInYears/100;
        return price - (discountForLoyaltyInPercentage * price);
    }

    public static ApplyDiscount(price: number, accountStatus: AccountStatus, timeOfHavingAccountInYears: number): number
    {
        let priceAfterDiscount: number = 0;
        switch (accountStatus)
        {
            case AccountStatus.NotRegistered:
                priceAfterDiscount = price;
                break;
            case AccountStatus.SimpleCustomer:
                priceAfterDiscount = DiscountManager.ApplyDiscountForAccountStatus(price, DiscountManager.DiscountForSimpleCustomers);
                break;
            case AccountStatus.ValuableCustomer:
                priceAfterDiscount = DiscountManager.ApplyDiscountForAccountStatus(price,  DiscountManager.DiscountForValuableCustomers);
                break;
            case AccountStatus.MostValuableCustomer:
                priceAfterDiscount = DiscountManager.ApplyDiscountForAccountStatus(price,  DiscountManager.DiscountForMostValuableCustomers);
                break;
            default:
                throw new Error('Not implementet');
        }
        priceAfterDiscount = DiscountManager.ApplyDiscountForTimeOfHavingAccount(priceAfterDiscount, timeOfHavingAccountInYears);
        return priceAfterDiscount;
    }
}